src/EditorFeatures/Core/Diagnostics/DiagnosticsClassificationTaggerProvider.cs

```diff
@@ -62,18 +61,61 @@ protected internal override bool IsEnabled
             => !_editorOptionsService.Factory.GlobalOptions.GetOptionValue(DefaultTextViewHostOptions.IsInContrastModeId);
 
         protected internal override bool IncludeDiagnostic(DiagnosticData data)
-            => data.CustomTags.Contains(WellKnownDiagnosticTags.Unnecessary);
+        {
+            // Fading diagnostic might have an 'Unnecessary' custom tag or
+            // unnecessary additional location indices.
+            if (!data.CustomTags.Contains(WellKnownDiagnosticTags.Unnecessary)
+                && !TryGetUnnecessaryLocationIndices(data, out _))
```

Comment:
just to check, but is it intentional this is an `&&` check and not an `||` check?

Reply:
Per my understanding, yes the `&&` is intentional

|Has unnecessary tag?|Has location indices?|Do we want to proceed?|Note|
|-------------------|--------------------|----------------------|----|
| True | True | Yes | This analyzer gives specific locations to fade out and it has unnecessary tag |
| True | False | Yes | This analyzer doesn't give specific location, so we fade out the primary location. The `\|\|` will break this scenario |
| False | True | Yes | This analyzer has specific locations to fade, so we do that. I think this didn't used to work previously, see https://github.com/dotnet/roslyn/pull/62562#discussion_r918937532. The `\|\|` will break this scenario |
| False | False | No | Nothing we can do |

@mavasani can confirm if my understanding is correct or not :)

Reply:
great comment.  would be handy to have in the code to clarify all those cases.

Reply:
Thanks for the details @Youssef1313, your understanding is perfectly correct. Let me add detailed comment.

Reply:
So, I ended up avoiding this complicated system. Now, all unnecessary code diagnostics *must* add `Unnecessary` custom tag to the descriptor, even the ones that encode unnecessary location indices in the property bag. The unnecessary classification tagger already has the check to first check for these encoded unnecessary indices in the property bag before falling back to fading the primary location, so this just needed me to fix the analyzers that were not adding `Unnecessary` custom tag to the descriptor when reporting the encoded unnecessary location diagnostics